dali/pipeline/graph/op_graph.h

```diff
@@ -0,0 +1,367 @@
+        name + "\" has no known source.");
+    return tensor_nodes_[it->second].producer;
+  }
+
+  /**
+   * @brief Checks if given Tensor already exists in the graph
+   */
+  DLL_PUBLIC inline bool TensorExists(const string &name) {
+    auto it = tensor_name_to_id_.find(name);
+    return it != tensor_name_to_id_.end();
+  }
+
+  /**
+   * @brief Returns the id of the op that produces the tensor with
+   * the given name.
+   */
+  DLL_PUBLIC inline OpNodeId TensorSourceID(const string &name) {
+    return TensorSourceMeta(name).node;
+  }
+
+  /**
+   * @brief Returns the output idx of the input tensor in
+   * its source.
+   */
+  DLL_PUBLIC inline Index TensorIdxInSource(const string &name) {
+    return TensorSourceMeta(name).index;
+  }
+
+  /**
+   * @brief Returns true if the tensor with the given name
+   * has a backend type that matches the calling type.
+   */
+  template <typename Backend>
+  DLL_PUBLIC bool TensorIsType(const string &name);
+
+  /**
+   * @brief Returns a vector of meta-data about the nodes that
+   * consume the tensor with the input name.
+   */
+  DLL_PUBLIC inline vector<TensorMeta> TensorConsumerMeta(const string &name) const {
+    auto it = tensor_name_to_id_.find(name);
+    if (it == tensor_name_to_id_.end()) {
+      // If we have no entries for this tensors consumers,
+      // we just return an empty vector
+      return vector<TensorMeta>{};
+    }
+    return tensor_nodes_[it->second].consumers;
+  }
+
+  /**
+   * @brief Helper function for saving graph to DOT file
+   */
+  DLL_PUBLIC void GenerateDOTFromGraph(std::ofstream& ofs, bool show_tensors, bool show_ids,
+                                       bool use_colors);
+
+  /**
+   * @brief Instantiates the operators based on OpSpecs in nodes
+   */
+  DLL_PUBLIC void InstantiateOperators();
+
+  /**
+   * @brief Save graph in DOT directed graph format
+   * in filename.
+   */
+  DLL_PUBLIC void SaveToDotFile(const string filename, bool show_tensors = false,
+                                bool show_ids = false, bool use_colors = false) {
+    std::ofstream ofs(filename);
+    visited_nodes_.clear();
+    ofs << "digraph graphname {\n";
+    GenerateDOTFromGraph(ofs, show_tensors, show_ids, use_colors);
+    ofs << "}\n";
+    visited_nodes_.clear();
+  }
+
+  DLL_PUBLIC std::vector<TensorNodeId> GetOutputs(const std::vector<string>& output_names);
+  DLL_PUBLIC std::vector<TensorNodeId> GetStageOutputs(OpType stage);
+
+  DISABLE_COPY_MOVE_ASSIGN(OpGraph);
+
+ private:
+  // Should be called only once for each tensor
+  void GenerateDOTFromGraph(const TensorNode& current_node, std::ofstream& ofs, bool show_tensors,
+                            bool show_ids);
+
+  void RepartitionOps();
+
+  OpNode& PlaceNewOp(OpType op_type, OpSpec op_spec, std::string instance_name);
+  TensorNode& PlaceNewTensor();
+
+
+  std::vector<OpNode> op_nodes_;
+  std::vector<TensorNode> tensor_nodes_;
+  std::vector<std::vector<OpNodeId>> op_paritions_;
```

Comment:
```suggestion
  std::vector<std::vector<OpNodeId>> op_partitions_;
```
Amazing, what code completion can propagate...

Reply:
Done :)